### PR TITLE
feat: Smoke Tests V2

### PR DESCRIPTION
**Pull request by sichanyoo**
Companion aws-sdk-swift PR: https://github.com/awslabs/aws-sdk-swift/pull/1791

## Issue \#

1779

## Description of changes

- Adds the smoke test code generator for generic Smithy spec, whose behaviors can be customized via method overrides in a child class.

## Scope

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/DirectedSwiftCodegen.kt [hunk 1]
```diff
@@ -94,6 +94,9 @@ class DirectedSwiftCodegen(val context: PluginContext) :
 
             integrations.forEach { it.writeAdditionalFiles(context, ctx, writers) }
 
+            LOGGER.info("[${service.id}] Generating smoke tests for service")
+            generateSmokeTests(ctx)
```
Comment (sichanyoo) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/DirectedSwiftCodegen.kt:98]: This is _the_ place where smoke test generation gets triggered

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HTTPBindingProtocolGenerator.kt [hunk 2]
```diff
@@ -390,6 +390,10 @@ abstract class HTTPBindingProtocolGenerator(
         }
     }
 
+    override fun generateSmokeTests(ctx: ProtocolGenerator.GenerationContext) {
+        return SmokeTestGenerator(ctx).generateSmokeTests()
+    }
```
Comment (sichanyoo) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/HTTPBindingProtocolGenerator.kt:395]: This concrete implementation generates smoke tests according to generic Smithy spec.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/ProtocolGenerator.kt [hunk 3]
```diff
@@ -128,6 +128,11 @@ interface ProtocolGenerator {
      */
     fun generateProtocolClient(ctx: GenerationContext)
 
+    /**
+     * Generate smoke tests defined on the service model
+     */
+    fun generateSmokeTests(ctx: GenerationContext)
```
Comment (sichanyoo) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/ProtocolGenerator.kt:134]: This gets implemented both in smithy-swift and aws-sdk-swift, each generating tests in generic Smithy way vs. AWS way, respectively.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt [hunk 4]
```diff
@@ -0,0 +1,267 @@
+package software.amazon.smithy.swift.codegen.integration
```
Comment (sichanyoo) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt:1]: Generates smoke tests according to generic Smithy spec. A child class in aws-sdk-swift extends this class to customize AWS specific logic by overriding vendor specific methods.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+    ) {
+        val result = if (isSuccess) "ok" else "not ok"
+        val error = if (errorExpected) "error expected from service" else "no error expected from service"
+        val skipped = if (isSkipped) " # skip" else ""
+        writer.write("print(\$S)", "$result $serviceName $operationName - $error$skipped")
+        if (printCaughtError) {
+            writer.write("print(\"# Caught unexpected error: \\(error)\")")
+        }
+        writer.write("return $isSuccess")
+    }
```
Comment (sichanyoo) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt:177]: Test result line produced here follows the [TAP protocol](https://testanything.org/tap-specification.html) as well as additional restrictions defined in SEPs.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        } else if (errorExpected) {
+            // If generic error was expected, print success and return true.
+            renderPrintTestResult(writer, true, serviceName, operationName, true)
+        } else {
+            // If expected success, print failure, the unexpected error, and return false.
+            renderPrintTestResult(writer, false, serviceName, operationName, false, printCaughtError = true)
+        }
+        writer.dedent()
+        writer.write("}")
+    }
```
Comment (sichanyoo) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt:241]: `renderDoCatchBlock()` contains the logic for test result & error handling codegen.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+
+    /**
+     * Default behavior is no-op; override this method for vendor-specific behavior.
+     * Implementation should set config fields using values from vendorParams or custom fileprivate variables.
+     */
+    open fun handleVendorParams(vendorParams: ObjectNode, writer: SwiftWriter) {
+        // Pseudo-code example:
+        //      writer.write("config.value1 = ${value1-extracted-from-vendorParams}")
+    }
```
Comment (sichanyoo) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt:254]: The vendor params are custom client config fields included for a given service client. In our case, all the config fields added to service clients in aws-sdk-swift.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/swiftmodules/ClientRuntimeTypes.kt [hunk 8]
```diff
@@ -91,6 +91,7 @@ object ClientRuntimeTypes {
         val InterceptorProvider = runtimeSymbol("InterceptorProvider", SwiftDeclaration.PROTOCOL)
```
Comment (sichanyoo) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/swiftmodules/ClientRuntimeTypes.kt:91]: This and below change are for test files that need to import these.

**Review by sichanyoo — COMMENTED** [review]
Comments for reviewers.

**Review by jbelkins — APPROVED** [review]
All of this is good.

I'm tempted to include a codegen switch to turn smoke test generation on/off, but as long as the tests are .gitignored and they remain reasonably small, I'm okay with generating them every time for now.

File: smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt [hunk 9]
```diff
@@ -0,0 +1,267 @@
...
+            // Exit with 0 or 1 based on allTestsPassed boolean.
+            renderExitBlock(writer)
+        }
+    }
+
+    //    if allTestsPassed {
+    //        exit(0)
+    //    } else {
+    //        exit(1)
+    //    }
```
Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt:119]: remove commented code

Comment (jbelkins) [inline smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/integration/SmokeTestGenerator.kt:119]: remove commented code

**Review by jbelkins — CHANGES_REQUESTED** [review]
one minor code cleanup

**Review by jbelkins — APPROVED** [review]

**Pull request merged**